Restore the MSVC port of OCaml

dra27 opened the pull request:
_None of this is yet ready for actual code review - in particular, quite a few commits intentionally make OCaml MSVC-only!_ It does, however, work, and I therefore think it's worth sharing, even if only to put it in the 5.1 milestone and to expose the task list publicly. Any and all questions about this very much welcomed, of course!

Last week, Microsoft added in [C11 atomics in Visual Studio 2022](https://devblogs.microsoft.com/cppblog/c11-atomics-in-visual-studio-2022-version-17-5-preview-2/) and this PR restores the MSVC port of OCaml on this basis. At the moment, 32-bit bytecode does not work, but I think this may be down a miscompilation in cl. 64-bit, however, is completely working and the testsuite on my machine in both bytecode and native code is fully passing.

I have disabled CI, with the intention of gradually re-enabling it as parts of this get carved out in separate PRs and the MSVC-only hacks get put into their proper places. In particular, rebasing and manually testing this PR should help to prevent further divergence affecting this port, even if it can't yet be fully re-enabled.

I've done a side-by-side disassembly of `amd64.o` and `amd64nt.obj`, so while the actual assembler file needs considerably tidying, I believe it is correct and that the remaining bug affecting native code is actually in ocamlopt, rather than the runtime.

- [x] `Iextcall` to non-alloc functions fixed (fairly sure this is affecting mingw-w64 in 5.0, but not as readily)
- [ ] MSVC atomics working as expected, rather than "as working"!
- [ ] 32-bit bytecode support (almost certainly MSVC bug(s)?)
- [ ] Macros for `__declspec(thread)` (possibly via C11 `threads.h`?)
- [ ] winpthreads integration by submodule (or possibly via C11 `threads.h`?)
- [ ] Eliminate C compiler warnings (fairly sure these are atomics bugs in MSVC)
- [ ] Formalise or eliminate the addition of `-nodefaultlib:LIBCMT`
- [ ] Fix the additional inclusions of `winbase.h` (affects mingw-w64 as well)
- [ ] Fix reliance on `sys/time.h` (affects mingw-w64 as well)
- [ ] Debugging parts (`-Od`, etc.) removed
- [ ] amd64nt.asm either eliminated (by using ocamlopt's DSL for the amd64 runtime support object) or updated to be slightly less-unmaintainable (either using MASM macros properly or switching the amd64.S to Intel syntax and generating all from a common base)
- [ ] Name-mangling properly fixed for MASM-compatibility
- [ ] Update `README.win32.adoc`

dra27 committed:
Explicitly compile the primitives object

In bytecode, when compiling a custom runtime (when using any of -custom,
-output-obj, -output-complete-obj or -output-complete-exe), the
camlprim.c file is passed directly to the linker. When this was
originally done, the cost was a single typedef for the value type, but
now this process involves considerable code duplication both for the
-fdebug-prefix-map implementation and for the full definition of the
value type.

The primitives file is now explicitly compiled, which means it gets
treated in the same way as a C file passed to ocamlc and in particular
can `#include <caml/mlvalues.h>` to remove the definitions otherwise
duplicated with <caml/config.h>. Using `Ccomp.compile_file` also allows
the duplicate machinery for `-fdebug-prefix-map` to be deleted from
Bytelink.

TODO: the duplication on guarded_primitives should be improved. In fact,
can that be converted to have a different header from mlvalues.h (i.e. a
subset header included by mlvalues.h?)

dra27 committed:
Update msvs-promote-path

Principally to deal with the deprecation warning on the fgrep command.
Synchronises with version 0.5.0 from msvs-tools.

dra27 committed:
Re-enable MSVC (bytecode-only)

dra27 committed:
__declspec(thread)

dra27 committed:
Add winpthreads (somewhat unconditionally)

dra27 committed:
Don't ever include unistd.h on Windows

Test this on mingw-w64 - should go into 5.0 as well

dra27 committed:
Can't build with -WX at present

dra27 committed:
Fix use of /link with cl

dra27 committed:
WIP: add -nodefault:LIBCMT

Not sure why this is suddenly necessary, but silences a linker warning
building runtime/ocamlruns.exe

dra27 committed:
WIP: Unexpected places where winbase.h needed

Similarly feeds back to the mingw-w64 port.

dra27 committed:
WIP: lack of sys/time.h

usleep is part of unistd - mingw-w64 emulation should be removed.

dra27 committed:
MSVC BUG(?): uninitialised variables (Warning C4700)

Without this diff, various initialised variable warnings.

dra27 committed:
MSVC BUG(?): direct assignment to atomics not working

This function segfaults on entry with those three pointers as direct
assignments.

dra27 committed:
Expose cflags variable in ocamltest

dra27 committed:
MSVC BUG(?): use _Atomic uintnat instead of pointers

The pointer versions aren't generating code with cl - hopefully
equivalent use of uintnat being cast back to pointers _after_ they've
been atomically retrieved.

dra27 committed:
MSVC BUG(?): highly suspect changes needed to build systhreads

Parse errors from pthread.h?

dra27 committed:
MSVC BUG(?): Suppress warnings to make lf_skiplist test compile

dra27 committed:
Use ${cflags} in output-complete-obj tests

Explicitly compiles the object using the C compiler (which inherits
internal flags) rather than passing it to flexlink.

This is likely to be a good change - passing a C file to flexlink
includes no flags whatsoever.

dra27 committed:
Temporarily disable optimisations in C compilation

dra27 committed:
Somewhat clunkily enable debug information

dra27 committed:
Revert "Somewhat clunkily enable debug information"

This reverts commit 0bb616685cd214a706a9c07bc27e29b2511f23df.

dra27 committed:
WIP: Restore the native backend

The translation of amd64nt.asm is very rough-and-ready, but it does
produce the same code as amd64.S does for mingw-w64.

Edit: at least it did the first, it should be checked again!

dra27 committed:
Reverse the name-mangling changes for MASM-compatibility

Might actually be possible to use $ in general??

dra27 committed:
Temporarily dump the assembly files

Enables "source" debugging (or at least, symbol de-mangling)

dra27 committed:
Temporary: disable pull request CI

nojb commented on runtime/platform.c:
```diff
@@ -209,23 +211,29 @@ void caml_mem_decommit(void* mem, uintnat size)
 #define Slow_sleep_ns    1000000 //  1 ms
 #define Max_sleep_ns  1000000000 //  1 s
 
+/* Again, TODO: how does this work on mingw-w64 atm? */
```
Mingw-w64 emulation is at:

https://github.com/mirror/mingw-w64/blob/master/mingw-w64-crt/misc/mingw_usleep.c

If I understand correctly, we would include some equivalent code to have it available for MSVC as well?

dra27 replied:
Thanks for digging! TIL there are many more emulated CRT functions in mingw-w64! Indeed, this looks like something we'd want to move to win32.c as `caml_usleep` and then have `osdeps.h` (I think?) deal with it.

In the absence of benchmarks, I'm not sure what's worse - `usleep` in mingw-w64 is actually "millisecond sleep", whereas I'm not sure that the function I put in platform.c (which I very quickly grabbed from somewhere!) making that many system calls is going to be particularly good at sub-ms granularity). On the other hand, this is used for `SPIN_WAIT` throughout the gc.

XVilka replied:
The most precise way to sleep on Windows is this: https://stackoverflow.com/questions/7827062/is-there-a-windows-equivalent-of-nanosleep

dra27 replied:
Thanks @XVilka. though the issue is not the _precision_ but the _granularity_.

dra27 committed:
Implement Caml_state in MASM

domain_state.tbl is pre-processed to remove comments and evaluate
conditionals. The result can be included directly in the backend files
for MASM which both simplifies the overall appearance of the code and
also eliminates the two awk scripts which were used previously.

dra27 committed:
MSVC BUG(?): additional workaround in codefrag.c

Need in the mark1 branch as well - possible regression in MSVC since
Dec 2022?

dra27 committed:
MSVC BUG(?): Another uintnat workaround

dra27 committed:
Install default.manifest on 32bit still

Slight over-removal in PR#11904

dra27 committed:
MSVC BUG(?): atomic_uintnat * incorrectly identified as atomic type

Rather than a non-atomic pointer to an atomic type.

dra27 committed:
Avoid pointer arithmetic over void *

dra27 committed:
Temporary: rename C stub test

Really should polish off masm-reserved from dra27/ocaml rather than
this...

dra27 committed:
WIP: runtime isn't linking with libpthread.lib at present

dra27 committed:
Add cpu_relax

dra27 commented:
This PR has run its course - the reviewable work continues in #12954!
